File: tests/ccsp-comparisons/pycsp/agents.py
import math, sys

from pycsp import *

n_cycles = 256

world_size = 5
world_area = world_size * world_size

loc_size = 4096
loc_area = loc_size * loc_size
loc_agents = 12
loc_max = (loc_size // 2) - 1
loc_min = - (loc_size // 2)

offsets = [(-1, -1), (0, -1), (1, -1), (-1, 0), (1, 0), (-1, 1), (0, 1), (1, 1), (0,0)]
offset_dict = {}

LOC_ENTER = 0
LOC_MOVE = 1
LOC_BORROW_INFO = 2
LOC_SHUTDOWN = 3

LOC_STAY_HERE = 0
LOC_GO_THERE = 1

VIEW_BORROW_VIEW = 0
VIEW_SHUTDOWN = 1

def build_offset_dict():
	d = 0
	for o in offsets:
		offset_dict[o]	= d
		d		= d + 1

def vector_add(a, b):
	return (a[0] + b[0], a[1] + b[1])
def vector_sub(a, b):
	return (a[0] - b[0], a[1] - b[1])
def make_nv(v):
	return (v[0] * loc_size, v[1] * loc_size)

def in_bounds(p):
	if (p[0] < loc_min) or (p[0] > loc_max):
		return False
	elif (p[1] < loc_min) or (p[1] > loc_max):
		return False
	else:
		return True

def dir_value(n):
	if n < loc_min:
		return -1
	elif n > loc_max:
		return 1
	else:
		return 0

def merge_agents(agent_list, v, agents):
	for (id, loc, pos) in agents.values():
		agent_list.append((id, loc, vector_add(pos, v)))

@process
def view(req_chan, search):
	running   = True
	cycle     = -1
	cache     = []
	resp_chan = One2OneChannel()
	while running:
		msg = req_chan.read()
		req = msg[0]

		if req == VIEW_BORROW_VIEW:
			resp = msg[1]
			req_cycle = msg[2]
			if req_cycle != cycle:
				cycle = req_cycle
				cache = []
				for (o, loc) in zip(offsets, search):
					nv = make_nv(o)
					loc.write((LOC_BORROW_INFO, resp_chan))
					agents = resp_chan.read()
					merge_agents(cache, nv, agents)
					resp_chan.write(agents)
			resp.write(cache)
			cache = resp.read()
		elif req == VIEW_SHUTDOWN:
			running = False

def redirect_agent(resp, neighbours, info):
	pos       = info[2]
	n_offset  = (dir_value(pos[0]), dir_value(pos[1]))
	direction = offset_dict[n_offset]
	neighbour = neighbours[direction]
	pos       = vector_sub(pos, make_nv(n_offset))
	resp.write((LOC_GO_THERE, neighbour, (info[0], info[1], pos)))

@process
def location(loc, req_chan, neighbours):
	agents = {}
	running	= True
	
	view_chan = Any2OneChannel("view" + str(loc))
	Spawn(view(view_chan, neighbours + [req_chan]))

	while running:
		# msg = (Request, ...)
		msg  = req_chan.read()
		req  = msg[0]

		if req == LOC_ENTER:
			# msg = (Request, Response, Info)
			# info = (ID, Location, (X, Y)
			resp    = msg[1]
			info	= msg[2]
			id	= info[0]
			if in_bounds(info[2]):
				info = (id, loc, info[2])
				agents[id] = info
				resp.write((LOC_STAY_HERE, info, view_chan))
			else:
				redirect_agent(resp, neighbours, info)
		elif req == LOC_MOVE:
			# msg = (Request, Response, ID, Vector)
			resp    = msg[1]
			id	= msg[2]
			vector	= msg[3]
			info	= agents[id]
			pos	= vector_add (info[2], vector)
			info	= (id, loc, pos)
			if in_bounds(pos):
				agents[id] = info
				resp.write((LOC_STAY_HERE, info, view_chan))
			else:
				agents.pop(id)
				redirect_agent(resp, neighbours, info)
		elif req == LOC_BORROW_INFO:
			# msg = (Request, Response)
			resp    = msg[1]
			resp.write(agents)
			agents = resp.read()
		elif req == LOC_SHUTDOWN:
			view_chan.write((VIEW_SHUTDOWN,))
			running = False

def quot(x, d):
	r = x / float(d)
	if r < 0.0:
		return int(math.ceil(r))
	else:
		return int(math.floor(r))

def a_sqrt(x, r):
	last_r = 0
	step   = 0
	while (last_r != r) and (step < 16):
		last_r = r
		r      = (r + (x // r)) >> 1
		step   = step + 1
	return r

def agent_start(cycle, info):
	print str(cycle) + " " + str(info[0]) + " start"
def agent_at(cycle, info):
	print str(cycle) + " " + str(info[0]) + " at " + str(info[1]) + ":" + str(info[2][0]) + "," + str(info[2][1])
def agent_end(cycle, info, persona):
	print str(cycle) + " " + str(info[0]) + " end " + str(persona)

@process
def agent(cycle, info, loc, bar):
	persona = info[0] * 37
	resp 	= One2OneChannel()

	agent_start(cycle, info)

	bar.sync()

	loc.write((LOC_ENTER, resp, info))
	msg  = resp.read()
	assert (msg[0] == LOC_STAY_HERE)
	info = msg[1]
	view = msg[2]

	bar.sync()

	while cycle < n_cycles:
		persona = (persona & 65535) * info[2][0]
		persona = (persona & 65535) * info[2][1]
		persona = (persona & 65535) * info[1]
		
		agent_at(cycle, info)
		
		cycle   = cycle + 1

		view.write((VIEW_BORROW_VIEW, resp, cycle))
		agents = resp.read()
		
		persona = persona + len(agents)
		px	= persona & 255
		py	= (persona >> 8) & 255
		force	= [0, 0]
		
		for agent in agents:
			if agent[0] == info[0]:
				continue

			d   = vector_sub(info[2], agent[2])
			dx  = quot((d[0] * (px + 1)), 128)
			dy  = quot((d[1] * (py + 1)), 128)
			dx2 = dx * dx
			dy2 = dy * dy
			r2  = dx2 + dy2
			f   = min(loc_size, (3 * loc_area) // (r2 + 1))
			r   = 1

			if dx2 > dy2:
				r = a_sqrt (r2, abs(dx))
			elif dy2 > 0:
				r = a_sqrt (r2, abs(dy))

			if r2 != 0:
				force[0] = force[0] + quot((f * dx), r)
				force[1] = force[1] + quot((f * dy), r)
			else:
				x = (cycle & 1) * (-1)
				y = ((cycle >> 1) & 1) * (-1)
				force[0] = force[0] + (f * x)
				force[1] = force[1] + (f * y)
		resp.write(agents)
		bar.sync()

		loc.write((LOC_MOVE, resp, info[0], tuple(force)))

		moving = True
		while moving:
			msg = resp.read()
			if msg[0] == LOC_STAY_HERE:
				info   = msg[1]
				view   = msg[2]
				moving = False
			elif msg[0] == LOC_GO_THERE:
				loc    = msg[1]
				info   = msg[2]
				loc.write((LOC_ENTER, resp, info))
		
		bar.sync()
	
	bar.sync()
	agent_at(cycle, info)
	agent_end(cycle, info, persona)

def neighbour_offset(loc, o):
	x = ((loc % world_size) + o[0]) % world_size
	y = ((loc // world_size) + o[1]) % world_size
	return (y * world_size) + x

def add_agents(id, loc, bar):
	for i in range(loc_agents):
		o = loc_min + ((loc_size // (loc_agents + 4)) * (i + 2))
		info = (id + i, -1, (o, o))
		Spawn(agent(0, info, loc, bar))
	return (id + loc_agents)

def agents():
	build_offset_dict()
	bar = Barrier((world_area * loc_agents) + 1)
	world = [ Any2OneChannel("loc" + str(i)) for i in range(world_area) ]
	id = 1
	for loc in range(world_area):
		neighbours = [ world[neighbour_offset(loc, o)] for o in offsets ]
		Spawn(location(loc, world[loc], neighbours))
		id = add_agents(id, world[loc], bar)
	for i in range(n_cycles + 1):
		bar.sync()
		bar.sync()
	bar.sync()
	for loc in range(world_area):
		world[loc].write((LOC_SHUTDOWN,))

if __name__ == "__main__":
	agents()

File: tests/ccsp-comparisons/ccsp/agents.c
#include <stdlib.h>
#include <stdio.h>
#include <assert.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <unistd.h>

#include <cif.h>

#define N_CYCLES	1024

#define WORLD_SIZE	5

#define LOC_SIZE	4096
#define	LOC_AREA	(LOC_SIZE * LOC_SIZE)
#define LOC_AGENTS	12

#define	LOC_MAX		((LOC_SIZE / 2) - 1)
#define LOC_MIN		(-(LOC_SIZE / 2))

typedef struct _vector_t {
	int	x;
	int	y;
} vector_t;

typedef struct _agent_t {
	int		id;
	int		loc;
	vector_t	pos;
} agent_t;

typedef struct _agent_elem_t {
	int		prev;
	int		next;
	agent_t		data;
} agent_elem_t;

typedef struct _agent_list_t {
	int		size;
	int		used;
	int		head;
	int		tail;
	int		free;
	agent_elem_t	elements[1];
} agent_list_t;

enum {
	LOC_ENTER,
	LOC_MOVE,
	LOC_GET_VIEW,
	LOC_BORROW_INFO,
	LOC_SHUTDOWN,
	LOC_STAY_HERE,
	LOC_GO_THERE
};

enum {
	VIEW_BORROW_VIEW,
	VIEW_SHUTDOWN
};

typedef struct _move_req_t {
	int		id;
	vector_t	v;
} move_req_t;

typedef struct _view_req_t {
	int		type;
	int		cycle;
} view_req_t;

#define NEIGHBOURS 8
static const vector_t offsets[] = {
	{ .x = -1, .y = -1 },
	{ .x =  0, .y = -1 },
	{ .x =  1, .y = -1 },
	{ .x = -1, .y =  0 },
	{ .x =  1, .y =  0 },
	{ .x = -1, .y =  1 },
	{ .x =  0, .y =  1 },
	{ .x =  1, .y =  1 },
	{ .x =  0, .y =  0 }
};

static int world_size = WORLD_SIZE;
static int loc_agents = LOC_AGENTS;

static inline void vector_add (vector_t *a, const vector_t *b)
{
	a->x += b->x;
	a->y += b->y;
}

static inline void vector_sub (vector_t *a, const vector_t *b)
{
	a->x -= b->x;
	a->y -= b->y;
}

static inline int dir_value (int n)
{
	if (n < LOC_MIN)
		return -1;
	else if (n > LOC_MAX)
		return 1;
	else
		return 0;
}

static inline int in_bounds (vector_t *p)
{
	if (p->x < LOC_MIN || p->x > LOC_MAX)
		return 0;
	else
		return !(p->y < LOC_MIN || p->y > LOC_MAX);
}

static void reset_agent_list (agent_list_t *l) {
	int i;
	
	l->used = 0;
	l->head = -1;
	l->tail = -1;
	l->free = 0;

	for (i = 0; i < l->size - 1; ++i)
		l->elements[i].next = i + 1;
	l->elements[l->size - 1].next = -1;
}

static agent_list_t *alloc_agent_list (Workspace wptr)
{
	agent_list_t *l = (agent_list_t *) MAlloc (wptr, sizeof (agent_list_t));

	l->size = 1;
	reset_agent_list (l);

	return l;
}

static agent_t *add_agent (Workspace wptr, agent_list_t **list, agent_t *info)
{
	agent_list_t *l = *list;
	int idx;

	if (l->free == -1) {
		int old_size = l->size;
		int new_size = 2 + (old_size - 1) + ((old_size - 1) / 2);
		int ob, nb;
		int i;

		nb = sizeof (agent_list_t) + (sizeof (agent_elem_t) * (new_size - 1));
		ob = sizeof (agent_list_t) + (sizeof (agent_elem_t) * (old_size - 1));

		l = (agent_list_t *) MAlloc (wptr, nb);
		memcpy (l, *list, ob);
		l->size = new_size;
		l->free = old_size;
		for (i = old_size; i < (new_size - 1); ++i) {
			l->elements[i].next = i + 1;
		}
		l->elements[new_size - 1].next = -1;
		MRelease (wptr, *list);
		*list = l;
	}

	assert (l->used < l->size);

	idx = l->free;
	l->used++;
	l->free = l->elements[idx].next;
	if (l->tail != -1)
		l->elements[l->tail].next = idx;
	l->elements[idx].prev = l->tail;
	l->elements[idx].next = -1;
	l->tail = idx;

	if (l->head == -1)
		l->head = idx;

	assert (l->used < l->size || (l->used == l->size && l->free == -1));
	
	memcpy (&(l->elements[idx].data), info, sizeof (*info));

	return &(l->elements[idx].data);
}

static void remove_agent (Workspace wptr, agent_list_t **list, int idx)
{
	agent_list_t *l = *list;
	int prev = l->elements[idx].prev;
	int next = l->elements[idx].next;

	if (prev != -1)
		l->elements[prev].next = next;
	if (next != -1)
		l->elements[next].prev = prev;
	if (l->head == idx)
		l->head = next;
	if (l->tail == idx)
		l->tail = prev;

	l->elements[idx].next = l->free;
	l->used--;
	l->free	= idx;
}

static int find_agent (agent_list_t *l, int id)
{
	int p = l->head;

	assert (p != -1);

	while (l->elements[p].data.id != id) {
		p = l->elements[p].next;
		assert (p != -1);
	}

	return p;
}

static void merge_agents (Workspace wptr, const vector_t *offset, agent_list_t **result, agent_list_t *loc)
{
	int p = loc->head;

	while (p != -1) {
		agent_t *info = add_agent (wptr, result, &(loc->elements[p].data));
		info->pos.x += offset->x * LOC_SIZE;
		info->pos.y += offset->y * LOC_SIZE;
		p = loc->elements[p].next;
	}
}

static void compile_view (Workspace wptr, mt_cb_t **search, agent_list_t **result)
{
	int i;

	reset_agent_list (*result);

	for (i = 0; i < (NEIGHBOURS + 1); ++i) {
		agent_list_t *agents = NULL;
		mt_cb_t *loc = search[i];

		MTLock (wptr, loc, MT_CB_CLIENT);
		ChanOutInt (wptr, &(loc->channels[0]), LOC_BORROW_INFO);
		MTChanIn (wptr, &(loc->channels[1]), (void *) &agents);
		merge_agents (wptr, &(offsets[i]), result, agents);
		MTChanOut (wptr, &(loc->channels[0]), (void *) &agents);
		MTUnlock (wptr, loc, MT_CB_CLIENT);
	}
}

#define VIEW_STACK 1024
static void view (Workspace wptr)
{
	void *fbar		= ProcGetParam (wptr, 2, void *);
	mt_array_t *sd 		= ProcGetParam (wptr, 1, mt_array_t *);
	mt_cb_t	*svr		= ProcGetParam (wptr, 0, mt_cb_t *);
	mt_cb_t **search	= (mt_cb_t **) sd->data;
	agent_list_t *data	= alloc_agent_list (wptr);
	Channel	*req		= &(svr->channels[0]);
	Channel *resp		= &(svr->channels[1]);
	int running 		= 1;
	int cycle 		= -1;

	while (running) {
		view_req_t msg;

		ChanIn (wptr, req, &msg, sizeof (msg));
		if (msg.type == VIEW_BORROW_VIEW) {
			if (cycle != msg.cycle) {
				compile_view (wptr, search, &data);
				cycle = msg.cycle;
			}
			MTChanOut (wptr, resp, (void *) &data);
			MTChanIn (wptr, req, (void *) &data);
		} else if (msg.type == VIEW_SHUTDOWN) {
			running = 0;
		}
	}

	MRelease (wptr, data);
	MTRelease (wptr, sd);
	MTRelease (wptr, svr);
	MTRelease (wptr, fbar);
}

static mt_array_t *build_search (Workspace wptr, mt_cb_t *self, mt_cb_t **neighbours)
{
	const int cb_type 	= MT_SIMPLE | MT_MAKE_TYPE (MT_CB) | MT_CB_SHARED;
	mt_array_t *sd		= MTAllocArray (wptr, cb_type, 1, (NEIGHBOURS + 1));
	mt_cb_t **search 	= (mt_cb_t **) sd->data;
	int i;
	
	for (i = 0; i < NEIGHBOURS; ++i) {
		search[i] = MTClone (wptr, neighbours[i]);
	}
	search[NEIGHBOURS] = MTClone (wptr, self);

	return sd;
}

static void redirect_agent (Workspace wptr, mt_cb_t **neighbours, Channel *resp, agent_t *info)
{
	vector_t o;
	int d = 0;

	o.x = dir_value (info->pos.x);
	o.y = dir_value (info->pos.y);
	while (offsets[d].x != o.x || offsets[d].y != o.y)
		d++;
	
	assert (d < NEIGHBOURS);

	o.x *= LOC_SIZE;
	o.y *= LOC_SIZE;

	vector_sub (&(info->pos), &o);

	ChanOutInt (wptr, resp, LOC_GO_THERE);
	ChanOut (wptr, resp, info, sizeof (*info));
	MTChanOut (wptr, resp, (void *) &(neighbours[d]));
}

#define LOCATION_STACK 1024
static void location (Workspace wptr)
{
	void 	*fbar	= ProcGetParam (wptr, 3, void *);
	mt_array_t *nd 	= ProcGetParam (wptr, 2, mt_array_t *);
	mt_cb_t	*svr	= ProcGetParam (wptr, 1, mt_cb_t *);
	int	loc	= ProcGetParam (wptr, 0, int);

	agent_list_t *state	= alloc_agent_list (wptr);
	mt_cb_t **neighbours	= (mt_cb_t **) nd->data;
	Channel	*req		= &(svr->channels[0]);
	Channel *resp		= &(svr->channels[1]);
	int running 		= 1;

	mt_cb_t	*view_cb	= MTAllocChanType (wptr, 2, true);
	
	{
		mt_array_t *search	= build_search (wptr, svr, neighbours);
		Workspace ws		= ProcAlloc (wptr, 3, VIEW_STACK);

		ProcMTCopy (wptr, ws, 0, view_cb);
		ProcMTMove (wptr, ws, 1, &search);
		ProcMTCopy (wptr, ws, 2, fbar);

		ProcStart (wptr, ws, view);
		
		MTRelease (wptr, view_cb); /* duplicate reference */
	}

	while (running) {
		int type;
		
		ChanInInt (wptr, req, &type);
		
		if (type == LOC_ENTER) {
			agent_t info;

			ChanIn (wptr, req, &info, sizeof (info));
			
			info.loc = loc;
			if (in_bounds (&(info.pos))) {
				ChanOutInt (wptr, resp, LOC_STAY_HERE);
				ChanOut (wptr, resp, &info, sizeof (info));
				add_agent (wptr, &state, &info);
			} else {
				redirect_agent (wptr, neighbours, resp, &info);
			}
		} else if (type == LOC_MOVE) {
			move_req_t update;
			agent_t *info;
			int idx;
			
			ChanIn (wptr, req, &update, sizeof (update));
			
			idx	= find_agent (state, update.id);
			info	= &(state->elements[idx].data);
			vector_add (&(info->pos), &(update.v));
			if (in_bounds (&(info->pos))) {
				ChanOutInt (wptr, resp, LOC_STAY_HERE);
				ChanOut (wptr, resp, info, sizeof (*info));
			} else {
				redirect_agent (wptr, neighbours, resp, info);
				remove_agent (wptr, &state, idx);
			}
		} else if (type == LOC_GET_VIEW) {
			MTChanOut (wptr, resp, (void *) &view_cb);
		} else if (type == LOC_BORROW_INFO) {
			MTChanOut (wptr, resp, (void *) &state);
			MTChanIn (wptr, req, (void *) &state);
		} else if (type == LOC_SHUTDOWN) {
			view_req_t vr;

			vr.type = VIEW_SHUTDOWN;
			MTLock (wptr, view_cb, MT_CB_CLIENT);
			ChanOut (wptr, &(view_cb->channels[0]), &vr, sizeof (vr));
			MTUnlock (wptr, view_cb, MT_CB_CLIENT);
			
			running = 0;
		}
	}

	MTRelease (wptr, view_cb);
	MTRelease (wptr, nd);
	MRelease (wptr, state);
	MTRelease (wptr, svr);
	MTRelease (wptr, fbar);
}

static int a_sqrt (const int x, int r)
{
	int last_r = 0;
	int step = 0;

	do {
		last_r = r;
		r = (r + (x / r)) >> 1;
		step++;
	} while (r != last_r && step < 16);

	return r;
}

#define AGENT_STACK 4096
static void agent (Workspace wptr)
{
	void	*fbar	= ProcGetParam (wptr, 5, void *);
	void 	*b	= ProcGetParam (wptr, 4, void *);
	mt_cb_t	*loc	= ProcGetParam (wptr, 3, mt_cb_t *);
	int	cycle	= 0;

	agent_t info;
	mt_cb_t *view = NULL;
	int persona, resp;

	info.id	 	= ProcGetParam (wptr, 0, int);
	info.loc	= -1;
	info.pos.x	= ProcGetParam (wptr, 1, int);
	info.pos.y	= ProcGetParam (wptr, 2, int);
	persona		= info.id * 37;

	printf ("%d %d start\n", cycle, info.id);

	MTSync (wptr, b);

	MTLock (wptr, loc, MT_CB_CLIENT);
	ChanOutInt (wptr, &(loc->channels[0]), LOC_ENTER);
	ChanOut (wptr, &(loc->channels[0]), &info, sizeof (info));
	ChanInInt (wptr, &(loc->channels[1]), &resp);
	assert (resp == LOC_STAY_HERE);
	ChanIn (wptr, &(loc->channels[1]), &info, sizeof (info));
	ChanOutInt (wptr, &(loc->channels[0]), LOC_GET_VIEW);
	MTChanIn (wptr, &(loc->channels[1]), (void *) &view);
	MTUnlock (wptr, loc, MT_CB_CLIENT);

	MTSync (wptr, b);

	while (cycle < N_CYCLES) {
		agent_list_t *agents = NULL;
		move_req_t mr;
		view_req_t vr;
		int idx, px, py;
		
		mr.id	= info.id;
		mr.v.x	= 0;
		mr.v.y	= 0;

		persona = (persona & 65535) * info.pos.x;
		persona = (persona & 65535) * info.pos.y;
		persona = (persona & 65535) * info.loc;

		//printf ("%d %d at %d:%d,%d\n", cycle, info.id, info.loc, info.pos.x, info.pos.y);
		
		cycle++;

		vr.type	= VIEW_BORROW_VIEW;
		vr.cycle = cycle;

		MTLock (wptr, view, MT_CB_CLIENT);
		ChanOut (wptr, &(view->channels[0]), &vr, sizeof (vr));
		MTChanIn (wptr, &(view->channels[1]), (void *) &agents);

		persona += agents->used;
		px	= persona & 0xff;
		py	= (persona >> 8) & 0xff;
		idx 	= agents->head;
		while (idx != -1) {
			agent_t *a_info = &(agents->elements[idx].data);
			int dx, dy, dx2, dy2, r2, r, f;
			idx		= agents->elements[idx].next;

			if (a_info->id == info.id)
				continue;

			dx = info.pos.x - a_info->pos.x;
			dy = info.pos.y - a_info->pos.y;
			dx = (dx * (px + 1)) / 128;
			dy = (dy * (py + 1)) / 128;
			dx2 = dx * dx;
			dy2 = dy * dy;
			r2 = dx2 + dy2;
			f = (3 * LOC_AREA) / (r2 + 1);
			if (f > LOC_SIZE)
				f = LOC_SIZE;
			
			if (dx2 > dy2)
				r = a_sqrt (r2, abs (dx));
			else if (dy2 > 0)
				r = a_sqrt (r2, abs (dy));
			else
				r = 1;

			if (r2 != 0) {
				mr.v.x += (f * dx) / r;
				mr.v.y += (f * dy) / r;
			} else {
				int x = (cycle & 1) * (-1);
				int y = ((cycle >> 1) & 1) * (-1);
				mr.v.x += f * x;
				mr.v.y += f * y;
			}
			
			Reschedule (wptr);
		}

		MTChanOut (wptr, &(view->channels[0]), (void *) &agents);
		MTUnlock (wptr, view, MT_CB_CLIENT);

		MTSync (wptr, b);

		idx = 1;
		do {
			MTLock (wptr, loc, MT_CB_CLIENT);
			if (idx == 1) {
				ChanOutInt (wptr, &(loc->channels[0]), LOC_MOVE);
				ChanOut (wptr, &(loc->channels[0]), &mr, sizeof (mr));
			} else {
				ChanOutInt (wptr, &(loc->channels[0]), LOC_ENTER);
				ChanOut (wptr, &(loc->channels[0]), &info, sizeof (info));
			}
			ChanInInt (wptr, &(loc->channels[1]), &resp);
			if (resp == LOC_STAY_HERE) {
				ChanIn (wptr, &(loc->channels[1]), &info, sizeof (info));
				if (idx > 1) {
					MTRelease (wptr, view);
					view = NULL;
					ChanOutInt (wptr, &(loc->channels[0]), LOC_GET_VIEW);
					MTChanIn (wptr, &(loc->channels[1]), (void *) &view);
				}
				MTUnlock (wptr, loc, MT_CB_CLIENT);
				idx = 0;
			} else if (resp == LOC_GO_THERE) {
				mt_cb_t *new_loc = NULL;
				ChanIn (wptr, &(loc->channels[1]), &info, sizeof (info));
				MTChanIn (wptr, &(loc->channels[1]), (void *) &new_loc);
				MTUnlock (wptr, loc, MT_CB_CLIENT);
				MTRelease (wptr, loc);
				loc = new_loc;
				idx++;
			}
		} while (idx);
		
		MTSync (wptr, b);
	}

	MTSync (wptr, b);

	printf ("%d %d at %d:%d,%d\n", cycle, info.id, info.loc, info.pos.x, info.pos.y);
	printf ("%d %d end %d\n", cycle, info.id, persona);

	MTRelease (wptr, view);
	MTRelease (wptr, loc);
	MTRelease (wptr, b);
	MTRelease (wptr, fbar);
}

static int wrap (int x, int max)
{
	while (x < 0)
		x += max;
	return x % max;
}

static mt_array_t *build_neighbours (Workspace wptr, int loc, mt_cb_t **world)
{
	const int cb_type = MT_SIMPLE | MT_MAKE_TYPE (MT_CB) | MT_CB_SHARED;
	mt_array_t *nd	= MTAllocArray (wptr, cb_type, 1, NEIGHBOURS);
	mt_cb_t **na	= (mt_cb_t **) nd->data;
	int x		= loc % world_size;
	int y		= loc / world_size;
	int i;

	for (i = 0; i < NEIGHBOURS; ++i) {
		int n_x = wrap (x + offsets[i].x, world_size);
		int n_y	= wrap (y + offsets[i].y, world_size);
		int n_loc = (n_y * world_size) + n_x;

		na[i] = (mt_cb_t *) MTClone (wptr, world[n_loc]);
	}

	return nd;
}

static void proc_main (Workspace wptr)
{
	const int cb_type 	= MT_SIMPLE | MT_MAKE_TYPE (MT_CB) | MT_CB_SHARED;
	int world_area		= world_size * world_size;
	mt_array_t *world_data	= MTAllocArray (wptr, cb_type, 1, world_area);
	mt_cb_t **world		= (mt_cb_t **) world_data->data;
	void *b, *afbar, *lfbar;
	int i, j, id;

	b	= MTAlloc (wptr, MT_MAKE_BARRIER (MT_BARRIER_FULL), 0);
	afbar	= MTAlloc (wptr, MT_MAKE_BARRIER (MT_BARRIER_FORKING), 0);
	lfbar	= MTAlloc (wptr, MT_MAKE_BARRIER (MT_BARRIER_FORKING), 0);

	for (i = 0; i < world_area; ++i)
		world[i] = MTAllocChanType (wptr, 2, true);
	
	for (i = 0; i < world_area; ++i) {
		mt_array_t *neighbours = build_neighbours (wptr, i, world);
		Workspace ws = ProcAlloc (wptr, 4, LOCATION_STACK);

		ProcParam (wptr, ws, 0, i);
		ProcMTCopy (wptr, ws, 1, world[i]);
		ProcMTMove (wptr, ws, 2, &neighbours);
		ProcMTCopy (wptr, ws, 3, lfbar);
		ProcStart (wptr, ws, location);
	}

	for (i = 0, id = 1; i < world_area; ++i) {
		for (j = 0; j < loc_agents; ++j, ++id) {
			int o = LOC_MIN + ((LOC_SIZE / (loc_agents + 4)) * (j + 2));
			Workspace ws = ProcAlloc (wptr, 6, AGENT_STACK);

			ProcParam (wptr, ws, 0, id);
			ProcParam (wptr, ws, 1, o);
			ProcParam (wptr, ws, 2, o);
			ProcMTCopy (wptr, ws, 3, world[i]);
			ProcMTCopy (wptr, ws, 4, b);
			ProcMTCopy (wptr, ws, 5, afbar);

			ProcStart (wptr, ws, agent);
		}
	}

	MTRelease (wptr, b);
	MTSync (wptr, afbar);
	
	for (i = 0; i < world_area; ++i) {
		MTLock (wptr, world[i], MT_CB_CLIENT);
		ChanOutInt (wptr, &(world[i]->channels[0]), LOC_SHUTDOWN);
		MTUnlock (wptr, world[i], MT_CB_CLIENT);
	}

	MTRelease (wptr, world_data);
	MTSync (wptr, lfbar);

	fflush (stdout);

	Shutdown (wptr);
}

int main (int argc, char *argv[])
{
	Workspace p;

	if (argc >= 2)
		world_size = atoi (argv[1]);
	if (argc >= 3)
		loc_agents = atoi (argv[2]);

	if (!ccsp_init ())
		return 1;

	p = ProcAllocInitial (0, 1024 * 1024);
	ProcStartInitial (p, proc_main);

	return 0;
}
